### PR TITLE
Hw 11 pecs

### PR DESCRIPTION
**Pull request by ihorromaniuk**

### CODE REVIEW

File: src/main/java/core/mate/academy/model/Bulldozer.java [hunk 1]
```diff
@@ -1,15 +1,39 @@
...
     public void doWork() {
         System.out.println("Bulldozer started to work");
     }
+
+    @Override
+
+    public List<Machine> get() {
+        Bulldozer oneCubeBulldozer = new Bulldozer(1);
+        Bulldozer oneAndHalfCubeBulldozer = new Bulldozer(1.5);
+        return List.of(oneCubeBulldozer, oneAndHalfCubeBulldozer);
```
Comment (vuyko23) [inline src/main/java/core/mate/academy/model/Bulldozer.java:29]: If you want you can use parents fild name,color in your constructor

**Review by vuyko23 — APPROVED** [review]

File: src/main/java/core/mate/academy/model/Machine.java [hunk 2]
```diff
@@ -1,6 +1,8 @@
 package core.mate.academy.model;
 
-public abstract class Machine implements Workable {
+import core.mate.academy.service.MachineProducer;
+
+public abstract class Machine implements Workable, MachineProducer {
```
Comment (KseniiaMakarova) [inline src/main/java/core/mate/academy/model/Machine.java:5]:
```suggestion
public abstract class Machine implements Workable {
```
you should create 3 Producer classes (1 for each type of machine) and override `get()` method there

File: src/main/java/core/mate/academy/service/MachineProducer.java [hunk 3]
```diff
@@ -1,11 +1,8 @@
...
 import java.util.List;
 
-public interface MachineProducer {
-    /**
-     * In your implementations - create 2-3 models, add them to the list and return
-     * @return - the list of models
-     */
-    List<Object> get();
+public interface MachineProducer<T extends Machine> {
+    List<Machine> get();
```
Comment (KseniiaMakarova) [inline src/main/java/core/mate/academy/service/MachineProducer.java:7]: let's make this method return List of any Machine (`List<Excavator>` or `List<Truck>`)

File: src/main/java/core/mate/academy/service/MachineService.java [hunk 4]
```diff
@@ -1,38 +1,12 @@
...
 
+import core.mate.academy.model.Machine;
 import java.util.List;
 
-/**
- * Parameterize this service and add its implementation in a separate class.
- */
-public interface MachineService {
+public interface MachineService<T extends Machine> {
+    List<Machine> getAll(Class<? extends Machine> type);
```
Comment (KseniiaMakarova) [inline src/main/java/core/mate/academy/service/MachineService.java:7]: in all method signatures of this class - let's not bind to the machines
you can operate `T` instead

File: src/main/java/core/mate/academy/service/MachineServiceImpl.java [hunk 5]
```diff
@@ -1,7 +1,36 @@
...
- */
-public class MachineServiceImpl {
+import core.mate.academy.model.Bulldozer;
+import core.mate.academy.model.Excavator;
+import core.mate.academy.model.Machine;
+import core.mate.academy.model.Truck;
+import java.util.ArrayList;
+import java.util.List;
+
+public class MachineServiceImpl<T extends Machine> implements MachineService<T> {
```
Comment (KseniiaMakarova) [inline src/main/java/core/mate/academy/service/MachineServiceImpl.java:10]:
```suggestion
public class MachineServiceImpl implements MachineService<Machine> {
```

File: src/main/java/core/mate/academy/service/MachineServiceImpl.java [hunk 6]
```diff
@@ -1,7 +1,36 @@
...
+    @Override
+    public List<Machine> getAll(Class<? extends Machine> type) {
+        Truck truck = new Truck();
+        Bulldozer bulldozer = new Bulldozer();
+        Excavator excavator = new Excavator();
+
+        return type.equals(Truck.class) ? truck.get()
+                : type.equals(Bulldozer.class) ? bulldozer.get()
```
Comment (KseniiaMakarova) [inline src/main/java/core/mate/academy/service/MachineServiceImpl.java:18]:
```suggestion
        return type.equals(Truck.class) ? new ArrayList(truckProducer.get())
                : type.equals(Bulldozer.class) ? new ArrayList(bulldozerProducer.get())
                ...
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/mate/academy/model/Truck.java [hunk 7]
```diff
@@ -5,11 +5,25 @@
  * Do not remove no-field constructor
  */
 public class Truck extends Machine {
+    private int wheelAmount;
```
Comment (KseniiaMakarova) [inline src/main/java/core/mate/academy/model/Truck.java:8]:
```suggestion
    private int numberOfWheels;
```

File: src/main/java/core/mate/academy/service/MachineService.java [hunk 8]
```diff
@@ -1,38 +1,12 @@
...
 
+import core.mate.academy.model.Machine;
 import java.util.List;
 
-/**
- * Parameterize this service and add its implementation in a separate class.
- */
-public interface MachineService {
+public interface MachineService<T extends Machine> {
+    List<T> getAll(Class<? extends Machine> type);
```
Comment (KseniiaMakarova) [inline src/main/java/core/mate/academy/service/MachineService.java:7]: let's use `T` everywhere

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
good job!

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**